### PR TITLE
Adds SQL Search Parameter Status Registry

### PR DESCRIPTION
**Pull request by rotodd**
## Description
- Adds a SQL table for the search parameter status registry (note: this will be merged with the existing search parameter table in [#AB74802](https://microsofthealth.visualstudio.com/Health/_workitems/edit/74802))
- Implements SQL stored procedure for getting and upserting search parameter statuses
- Initializes the search parameter status registry in `SqlServerFhirModel.cs`
- Renames status registry files for consistency:
  - `ISearchParameterRegistry` --> `IStatusRegistryDataStore`
  - `FilebasedSearchParameterRegistry` --> `FilebasedStatusRegistryDataStore`
  - `CosmosDbStatusRegistry` --> `CosmosDbStatusRegistryDataStore`

image

## Related issues
Addresses [#AB74236](https://microsofthealth.visualstudio.com/Health/_workitems/edit/74236) and [#AB74003](https://microsofthealth.visualstudio.com/Health/_workitems/edit/74003).

## Testing
- Adds integration tests for both the SQL and Cosmos DB search parameter status registry methods

### CODE REVIEW

File: test/Microsoft.Health.Fhir.Shared.Tests.Integration/Persistence/StatusRegistryDataStoreTests.cs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            }
+            finally
+            {
+                await _testHelper.DeleteSearchParameterStatusAsync(statusName1);
+                await _testHelper.DeleteSearchParameterStatusAsync(statusName2);
+            }
+        }
+
+        [Fact]
+        public async Task GivenAStatusRegistry_WhenUpsertingExistingStatuses_ThenTheStatusesAreNotAdded()
```
Comment (rbans96) [inline test/Microsoft.Health.Fhir.Shared.Tests.Integration/Persistence/StatusRegistryDataStoreTests.cs:76]: can we verify and rename this to ThenTheStatusesAreNotAdded -> ThenTheExistingStatusesAreUpdated

Comment (rotodd) [reply test/Microsoft.Health.Fhir.Shared.Tests.Integration/Persistence/StatusRegistryDataStoreTests.cs]: Good idea, thanks. Fixed in 94ea3cb.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/TvpRowGeneration/SearchParameterRegistryRowGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+{
+    internal class SearchParameterRegistryRowGenerator : ITableValuedParameterRowGenerator<List<ResourceSearchParameterStatus>, VLatest.SearchParamRegistryTableTypeRow>
+    {
+        public SearchParameterRegistryRowGenerator()
+        {
+        }
+
+        public IEnumerable<VLatest.SearchParamRegistryTableTypeRow> GenerateRows(List<ResourceSearchParameterStatus> searchParameterStatus)
+        {
+            return searchParameterStatus.Select(status => new VLatest.SearchParamRegistryTableTypeRow(status.Uri.ToString(), status.Status.ToString(), status.IsPartiallySupported)).ToList();
```
[the remark this answers is not in the document]
Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/TvpRowGeneration/SearchParameterRegistryRowGenerator.cs]: Fixed in 5e33a27.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/TvpRowGeneration/SearchParameterRegistryRowGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Storage.TvpRowGeneration
+{
+    internal class SearchParameterRegistryRowGenerator : ITableValuedParameterRowGenerator<List<ResourceSearchParameterStatus>, VLatest.SearchParamRegistryTableTypeRow>
+    {
+        public SearchParameterRegistryRowGenerator()
+        {
+        }
+
+        public IEnumerable<VLatest.SearchParamRegistryTableTypeRow> GenerateRows(List<ResourceSearchParameterStatus> searchParameterStatus)
```
[the remark this answers is not in the document]
Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/TvpRowGeneration/SearchParameterRegistryRowGenerator.cs]: Fixed in 5e33a27.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Storage/Registry/SqlServerStatusRegistryDataStore.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        {
+            EnsureArg.IsNotNull(statuses, nameof(statuses));
+
+            using (SqlConnectionWrapper sqlConnectionWrapper =
+                _sqlConnectionWrapperFactory.ObtainSqlConnectionWrapper(true))
+            using (SqlCommandWrapper sqlCommandWrapper = sqlConnectionWrapper.CreateSqlCommand())
+            {
+                VLatest.UpsertSearchParamStatus.PopulateCommand(sqlCommandWrapper, _updateSearchParamRegistryTvpGenerator.Generate(statuses.ToList()));
+
+                await sqlCommandWrapper.ExecuteScalarAsync(CancellationToken.None);
```
[the remark this answers is not in the document]
Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Storage/Registry/SqlServerStatusRegistryDataStore.cs]: Good catch. Fixed in 489a1a7.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/4.diff.sql [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    Uri varchar(128) COLLATE Latin1_General_100_CS_AS NOT NULL,
+    Status varchar(10) NOT NULL,
+    IsPartiallySupported bit NOT NULL
+)
+
+CREATE TABLE dbo.SearchParamRegistry
+(
+    Uri varchar(128) COLLATE Latin1_General_100_CS_AS NOT NULL,
+    Status varchar(10) NOT NULL,
+    LastUpdated datetimeoffset(7) NULL,
```
[the remark this answers is not in the document]
Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/4.diff.sql]: Nope, good point! Fixed in 8821804.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/4.diff.sql [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+--
+-- DESCRIPTION
+--     If a parameter with a matching URI already exists in the registry, it is updated.
+--     If not, a new entry is created.
+--
+-- PARAMETERS
+--     @searchParamStatuses
+--         * The updated or new search parameter statuses
+--
+CREATE PROCEDURE dbo.UpsertSearchParamStatus
```
[the remark this answers is not in the document]
Comment (rotodd) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/4.diff.sql]: Thanks, I think it's important to be consistent. Renamed in 8959737.

**Comment by rotodd** [comment]
Closing to create a new PR due to issues with keyvault in the test environment.

**Pull request closed without merging**